mesonbuild/msubprojects.py

```diff
@@ -248,21 +248,33 @@ def update_git(self):
             self.log(mlog.red(e.output))
             self.log(mlog.red(str(e)))
             return False
-        try:
-            # Fetch only the revision we need, this avoids fetching useless branches.
-            # revision can be either a branch, tag or commit id. In all cases we want
-            # FETCH_HEAD to be set to the desired commit and "git checkout <revision>"
-            # to to either switch to existing/new branch, or detach to tag/commit.
-            # It is more complicated than it first appear, see discussion there:
-            # https://github.com/mesonbuild/meson/pull/7723#discussion_r488816189.
-            heads_refmap = '+refs/heads/*:refs/remotes/origin/*'
-            tags_refmap = '+refs/tags/*:refs/tags/*'
-            self.git_output(['fetch', '--refmap', heads_refmap, '--refmap', tags_refmap, 'origin', revision])
-        except GitException as e:
-            self.log('  -> Could not fetch revision', mlog.bold(revision), 'in', mlog.bold(self.repo_dir))
-            self.log(mlog.red(e.output))
-            self.log(mlog.red(str(e)))
-            return False
+        if self.wrap_resolver.is_git_full_commit_id(revision) and \
+                quiet_git(['rev-parse', '--verify', revision + '^{commit}'], self.repo_dir)[0]:
+            # The revision we need is both a commit and available. So we do not
+            # need to fetch it because it cannot be updated.  Instead, trick
+            # git into setting FETCH_HEAD just in case, from the local commit.
+            self.git_output(['fetch', '.', revision])
+        else:
+            try:
+                # Fetch only the revision we need, this avoids fetching useless branches.
+                # revision can be either a branch, tag or commit id. In all cases we want
+                # FETCH_HEAD to be set to the desired commit and "git checkout <revision>"
+                # to to either switch to existing/new branch, or detach to tag/commit.
+                # It is more complicated than it first appear, see discussion there:
+                # https://github.com/mesonbuild/meson/pull/7723#discussion_r488816189.
+                heads_refmap = '+refs/heads/*:refs/remotes/origin/*'
+                tags_refmap = '+refs/tags/*:refs/tags/*'
+                self.git_output(['fetch', '--refmap', heads_refmap, '--refmap', tags_refmap, 'origin', revision])
+            except GitException as e:
+                self.log('  -> Could not fetch revision', mlog.bold(revision), 'in', mlog.bold(self.repo_dir))
+                if quiet_git(['rev-parse', revision + '^{commit}'], self.repo_dir)[0]:
+                    self.log(mlog.yellow('WARNING:'), 'Proceeding with locally available copy')
+                    # Trick git into setting FETCH_HEAD from the local revision.
+                    quiet_git(['fetch', '.', revision], self.repo_dir)
```

Comment:
Does this means that if the branch exists but it fails to fetch latest revision, that will still return success? I'm not sure this is a good idea, because you generally do want your CI script to fail on network issues. One particular case I have in mind is the current trend on renaming `master` branch to `main`, if one of my subproject does that, I want my CI to fail to update `master` branch unless of silently keep using old revision.

I don't really understand the point of calling `meson subprojects update` when you're offline at all, when does it make any sense?

Reply:
It makes sense in exactly the situation I described in the first place!

I don't understand why this is problematic for CI, it only succeeds if you side load a copy of the git repo so that the revision is satisfiable.

- It doesn't succeed on the first download of the subproject.
- It doesn't succeed if you eject the cache using subprojects purge.

Why does CI have a partial cache anyway? Its not available when your CI does `git clone` of the superproject sources.

Even if it was there (how???) and the CI has worse internet than my terrible home connection, the next git run will pick up the problem.

Or you could use pinned sources.

Reply:
If someone renames a long-lived branch then there will be lots of problems and nowhere near limited to meson. It will break git pull too, in mysterious and hard to fix ways.

But fully deleting the repo and re-cloning it will at least get ***you*** back to the renamed branch, and... if you use `revision = head` instead of naming the branch, it will fix meson too. So CI won't be affected, and users manually running these commands will have the log messages they need.

Reply:
@eli-schwartz GStreamer for example cache all subprojects in their docker image to only do an update when running the job. But IMHO it's just plain wrong to return success when it clearly was not a success and even printed a warning that it failed.